code/src/nuvla/job/job.py

```diff
@@ -41,6 +48,11 @@ def _init(self):
                 self.id = self.id.decode()
                 cimi_job = self.get_cimi_job(self.id)
                 dict.__init__(self, cimi_job)
+                if version_to_tuple(self.get('version')) > self._engine_version:
+                    log.warning(f"Job's version {self.get('version')} is higher than engine's {self._engine_version}. "
+                                "Putting job back to the queue.")
```

Comment:
in fact, if there's no other job in the queue, this will cause the job-executor to go into a noisy loop:
```
...
 | 2020-12-17 10:28:42,769 - WARNING - job.py:53 - Job's version 3 is higher than engine's 2.10.1-SNAPSHOT. Putting job back to the queue.
 | 2020-12-17 10:28:42,787 - WARNING - job.py:53 - Job's version 3 is higher than engine's 2.10.1-SNAPSHOT. Putting job back to the queue.
 | 2020-12-17 10:28:42,803 - WARNING - job.py:53 - Job's version 3 is higher than engine's 2.10.1-SNAPSHOT. Putting job back to the queue.
 | 2020-12-17 10:28:42,831 - WARNING - job.py:53 - Job's version 3 is higher than engine's 2.10.1-SNAPSHOT. Putting job back to the queue.
 | 2020-12-17 10:28:42,850 - WARNING - job.py:53 - Job's version 3 is higher than engine's 2.10.1-SNAPSHOT. Putting job back to the queue.
 | 2020-12-17 10:28:42,863 - WARNING - job.py:53 - Job's version 3 is higher than engine's 2.10.1-SNAPSHOT. Putting job back to the queue.
 | 2020-12-17 10:28:42,877 - WARNING - job.py:53 - Job's version 3 is higher than engine's 2.10.1-SNAPSHOT. Putting job back to the queue.
 | 2020-12-17 10:28:42,890 - WARNING - job.py:53 - Job's version 3 is higher than engine's 2.10.1-SNAPSHOT. Putting job back to the queue.
 | 2020-12-17 10:28:42,911 - WARNING - job.py:53 - Job's version 3 is higher than engine's 2.10.1-SNAPSHOT. Putting job back to the queue.
 | 2020-12-17 10:28:42,925 - WARNING - job.py:53 - Job's version 3 is higher than engine's 2.10.1-SNAPSHOT. Putting job back to the queue.
 | 2020-12-17 10:28:42,939 - WARNING - job.py:53 - Job's version 3 is higher than engine's 2.10.1-SNAPSHOT. Putting job back to the queue.
 | 2020-12-17 10:28:42,962 - WARNING - job.py:53 - Job's version 3 is higher than engine's 2.10.1-SNAPSHOT. Putting job back to the queue.
...
```

Reply:
30mil loop :) that's crazy. Agreed. 
First, I will change the log level to `DEBUG`. 
Regarding editing: we will hummer the server with too many edits. And I hope the job is not editable apart from its state and the output is the result of the execution, not a comment on scheduling issues. 
We can have a temporary (say 1 min) blacklist for job ids on the Executor. But the Job is stateless. We need to throw JobVersionTooOld with ID from Job. Executor would maintain the temp backlist and provide it next time to `Job(api, queue, bl=None)`.

Reply:
tbh, I'd rather just move the job to failed directly. It is very unlikely that during its lifetime the job-engine will get updated, so in most cases the job will simply timeout. 

updating the `ooutput` is nice, but still, without a failure, the user will not get notified, and will never see an end to his job. At least if the job fails right of the bet, the user will know the reason and fix it with a new job

Reply:
The comments are addressed in https://github.com/nuvla/job-engine/pull/149/commits/9805c082ac95a45f38389c5c6a24c3097fe97297